### DISCUSSION TITLE
"Extending a field plugin" Documentation

### DISCUSSION THREAD

**Discussion by 3lijah** [tinacms/tinacms#2765] [General]
Hello!

When we launched our site with TinaCMS in November 2021, we used a fairly messy/hacky way of using the GroupList plugin in the _app.js file and `<TinaCMS>` component.

I decided to try to pop back in and have a look to see if I could clean this up. I see that the latest version of TinaCMS has a much cleaner pattern and it even looks like the GroupList can now do what I was trying to hack in (getting the correct object title to appear in the list view).

After going through the documentation, I'm struggling to implement this with confidence.

Some examples:

https://tina.io/docs/advanced/extending-field-plugin/
- The "See here" link for the list of built-in component doesn't work.
- It shows how the validate function works, but doesn't document what the `allValues`, `meta`, or `fields` parameters do or how they're formatted.
- What is the `__type` key value for? Is it always just `'field'`?
- How do you use the `parse` and `format` functions?
- Later in the page it shows how to register the plugin. Where does `defineConfig` get imported from?

https://tina.io/docs/advanced/creating-field-component/
- The react-final-form docs are extensive. It would be nice to have a brief explanation on what the input and meta props look like are for without having to learn about react-final-form.
-  It would be great to have a full example that shows the component, its configuration, styling, and how it's fetched. I final it hard to map everything together when every example is different.

https://tina.io/docs/reference/toolkit/fields/group-list/
- It would be great to have an example on how this can be fully implemented.

Basically, I want to extend the GroupList . I feel like I have a few splotches of how to do it, but I'd love more guidance on how to put it all together.

**Comment by jeffsee55** [comment]
@3lijah these are really good questions, and we're in the process of redefining how to best extend Tina. 

> The react-final-form docs are extensive.

I think we'll likely start to restrict some parts of the raw final form APIs and wrap them up with something a little more polished, `final-form` is a lower level tool that probably shouldn't be leaking out to the public API. But in general Tina used to not have any knowledge of the `type` of a field, so `parse` and `format` could be used to cast values to `Number` or `Boolean` from strings, for example. So not as valuable today as they used to be.

In the short term, the work we've just done on "extending Tina" will mean we can remove some of the docs you're referencing.

> Basically, I want to extend the GroupList . I feel like I have a few splotches of how to do it, but I'd love more guidance on how to put it all together.

One of the problems with extending `GroupList` is that it's got all sorts of "list"-related capabilities that make it difficult to extend. We have drag-n-drop context and things like that which you'd have to reimplement. Our hope is that we can allow extensions to just plug in to the portions of the UI that devs actually care about. So having said that, what do you have in mind for extending this field?
